#### M4W Series

* [Table of Contents](M4WTOC.ipynb)
* <a href="https://colab.research.google.com/github/4dsolutions/m4w/blob/main/Polyhedrons.ipynb"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open and Execute in Google Colaboratory"></a>
* [![nbviewer](https://raw.githubusercontent.com/jupyter/design/master/logos/Badges/nbviewer_badge.svg)](https://nbviewer.org/github/4dsolutions/m4w/blob/main/Polyhedrons.ipynb)

# Polyhedrons &#x42F; Objects

<a data-flickr-embed="true" href="https://www.flickr.com/photos/kirbyurner/1023111093/" title="Concentric Hierarchy"><img src="https://live.staticflickr.com/1313/1023111093_6400ff79e8.jpg" width="500" height="375" alt="Concentric Hierarchy"/></a><script async src="//embedr.flickr.com/assets/client-code.js" charset="utf-8"></script>

One of our signature moves is to present object-oriented thinking (as in computer science), with polyhedrons as paradigm objects. With the Polyhedron type come its methods for rotating and scaling itself about the origin.

We start with a canonical arrangement of polyhedrons known to students of *Synergetics* as the concentric and or cosmic hierarchy. The accompanying volumes table centers around the unit edge tetrahedron as unit volume concept, and comes with a consistent vector algebra akin to XYZ's:  [Quadrays](https://github.com/4dsolutions/Python5/blob/master/Polyhedrons.ipynb).

Each polyhedron in the hierarchy (Tetrahedron, Cube, Octahedron...) is a subclass of Polyhedron, but with specific faces, tuples of vertices, hard-coded or selected from a database.

Verboten Math (Martian Math) Volumes Table:<br/>
<a data-flickr-embed="true" href="https://www.flickr.com/photos/kirbyurner/51846542476/" title="Concentric Hierarchy Volumes Table"><img src="https://live.staticflickr.com/65535/51846542476_653580f122.jpg" width="500" height="462" alt="Concentric Hierarchy Volumes Table"/></a><script async src="//embedr.flickr.com/assets/client-code.js" charset="utf-8"></script>

From *Synergetics*:<br/>
<a data-flickr-embed="true" href="https://www.flickr.com/photos/kirbyurner/4147716381/" title="Legendary Concentric Hierarchy"><img src="https://live.staticflickr.com/2573/4147716381_7991c273b2.jpg" width="389" height="500" alt="Legendary Concentric Hierarchy"/></a><script async src="//embedr.flickr.com/assets/client-code.js" charset="utf-8"></script>

Think Kepler

## Supermarket Math

To get the full experience of a database behind a web-based API, we store our [Polyhedrons](https://github.com/4dsolutions/Python5/blob/master/Polyhedrons%20101.ipynb) as related SQL tables and serve them out in response to HTTP requests. 

Starting with [SQLite](https://www.sqlite.org/index.html) inside of Python makes it easy, as no separate database engine need be started up or maintained. [`sqlite3`](https://docs.python.org/3/library/sqlite3.html) is already included in the [Python Standard Library](https://docs.python.org/3/tutorial/stdlib.html).

A database is where we keep track of inventory, save stuff, bank.

A typical table structure: a main table has one line per polyhedron, which points one-to-many to its faces, which in turn point one-to-many to the vertex vectors each face has, going around.  Edges distill from faces, e.g. if we have face ('A', 'B', 'C') then we have edges 'AB', 'BC', 'AC'.  To keep edge names unique, we tend to alphabetize them, such that 'CA' and 'AC' don't register separately. Just use 'AC'.

* Main --> each Polyhedron (name, number of vertices, edges, faces)
* Faces --> Edges distill on the fly from Faces
* Vertexes --> XYZ and/or Quadray coordinates, canonical positions

See my [Digital Mathematics from the Silicon Forest](https://wikieducator.org/Digital_Math) for more context (with Supermarket, Martian, Neolithic and Casino Math).

![hierarchy](https://upload.wikimedia.org/wikipedia/commons/d/dc/Povlabels.gif)

## Python Warmup

About `*` and `**` as "splat apart" vs "smash together" operators respectively

In [1]:
f = lambda *x: x        # combine as a tuple

arg  = ([1, 2], [3,4])  # two lists inside a tuple

f(*arg)                 # splat apart tuple then smash lists together

([1, 2], [3, 4])

## Polyhedron Play

In this section we're doing a closed loop transformation involving what David Koski and I have dubbed the "S factor", meaning the volumetric ratio of the S to E modules.

Another way to compute the same sfactor is to ratio the cuboctahedron with the icosahedron of same edges, the two being related by the Jitterbug Transformation.

The sfactor, applied to the cuboctahedron (CO) of volume 2.5 (half the edges of the edge 2R CO, R = IVM ball radius) may be modeled as the beginning of a twisting action, akin to the Jitterbug, but wherein edges are allowed to lengthen. 

The trianglar faces twist in a way that keeps them flush with a containing octahedron of volume 4. Two applications of the sfactor take us to the icosahedron with eight of its faces flush to said Octahedron. 

<a data-flickr-embed="true" href="https://www.flickr.com/photos/kirbyurner/53513118638/in/dateposted/" title="Screen Shot 2024-02-06 at 9.28.31 AM"><img src="https://live.staticflickr.com/65535/53513118638_9744a9b926_o.png" width="254" height="262" alt="Screen Shot 2024-02-06 at 9.28.31 AM"/></a><script async src="//embedr.flickr.com/assets/client-code.js" charset="utf-8"></script>
<br />
**From: RBF, Synergetics, Fig. 988.00 Polyhedral Evolution: S Quanta Module: Comparisons of skew polyhedra.**

This relationship of the two polys (octa and icosa) begets the S modules in the first place: 24 of them (12L and 12R) brick in the difference. 24 S modules, applied to the icosahedron, build its containing volume 4 octahedron.

We call this faces-flush icosahedron the IcosaWithin. The 2.5 CO morphs into the IcosaWithin with two applications of the sfactor i.e. one application of the 2nd power of the sfactor.

This IcosaWithin has edges equal to the sfactor. Multiplying by the reciprocal of the sfactor therefore models shrinking the IcosaWithin to one of edges 1R, i.e. 1/8th the volume of the Jitterbug Icosa, the one with edges equal the volume 20 CO's.

But those edges 1 match the those of the 2.5 volumed CO. One more application of the sfactor will therefore bring us full circle back to 2.5.  The we may start the whole cycle again:  2.5 CO times sfactor times sfactor brings us to IcosaWithin, which times 1/sfactor brings us to 1R edged Icosa, times the sfactor again brings us back to the 2.5 volumed CO. 

In [2]:
import sympy as sy
from sympy import Symbol, symbols

In [3]:
phi = (sy.sqrt(5)+1)/2

In [4]:
phi.evalf()

1.61803398874989

<a data-flickr-embed="true" href="https://www.flickr.com/photos/kirbyurner/53513018333/in/dateposted/" title="Screen Shot 2024-02-06 at 8.27.30 AM"><img src="https://live.staticflickr.com/65535/53513018333_80259f3324.jpg" width="500" height="272" alt="Screen Shot 2024-02-06 at 8.27.30 AM"/></a><script async src="//embedr.flickr.com/assets/client-code.js" charset="utf-8"></script>


In [5]:
icosavol = 5 * sy.sqrt(2) * phi**2  # icosa of edges 2R
icosavol.evalf()

18.5122958682192

In [6]:
sfactor = sy.Integer(20)/icosavol # CO:Icosa ratio = S:E ratio
sfactor = sfactor.simplify()
sfactor

-sqrt(10) + 3*sqrt(2)

In [7]:
sfactor.evalf()

1.08036302695091

![](http://rwgrayprojects.com/synergetics/s09/figs/f8813c.gif)

![](http://rwgrayprojects.com/synergetics/s09/figs/f8813a.gif)

In [8]:
a = Symbol('a')
a = 2
FE = a * sy.sqrt(7 - 3 * sy.sqrt(5))
FE.evalf() # also sfactor

1.08036302695091

$$
S Factor = 2\sqrt{7 - 3\sqrt{5}}
$$

In [9]:
icosa_xyz = a**3 * (5/12)*(3 + sy.sqrt(5)) # textbook XYZ volume

In [10]:
icosa_xyz.evalf()

17.4535599249993

In [11]:
S3 = sy.sqrt((9/8)) # Synergetics constant

In [12]:
S3

1.06066017177982

In [13]:
S3 * icosa_xyz.evalf() # convert to tetravolumes

18.5122958682192

In [14]:
IcosaWithin = sy.Rational(5,2) * sfactor**2 # 2.5 CO to IcosaWithin
Smod = (4 - IcosaWithin)/24
Smod = Smod.simplify()

In [15]:
Smod

-11/4 + 5*sqrt(5)/4

In [16]:
(24 * Smod + IcosaWithin).evalf()

4.00000000000000

In [17]:
sfactor.evalf()

1.08036302695091

In [18]:
tinyme = (icosavol/8).evalf(40) # 1/8th icosa of edges 2R, edges R
tinyme

2.314036983527395149501237411581816490446

Volume changes as a 3rd power of the scale factor as applied to edges. For example if we apply 1/sfactor then the corresponding volume changes by a 3rd power of 1/sfactor. This is only if we assume the shape is held constant, in terms of surface and central angles.

In [19]:
(8 * IcosaWithin * (1/sfactor**3)).evalf() # check

18.5122958682192

In [20]:
(IcosaWithin * 1/sfactor**3).evalf(40) # same as tinyme

2.314036983527395149501237411581816490446

In [21]:
(tinyme * sfactor).evalf(40) # Jitterbug from icosa to CO, back to 2.5

2.500000000000000000000000000000000000000

When we apply the sfactor to the 2.5 CO of edges 1R, we might imagine it going, right there and then, to a CO of edges sfactor and volume 2.5 times the sfactor to the 3rd power.

This very CO may then be shrunk to IcosaWithin with same length edges, and embedded in the Octa 4, by the application of 1/sfactor.

In [22]:
(2.5 * sfactor**3 * 1/sfactor).evalf() # IcosaWithin

2.91796067500631

David Koski:  Too cool Icosawithin is 60S + 20s3

What does he mean?  The S module scaled up by phi in linear dimensions, has its volume impacted by phi to the 3rd power. S3 (capital S) means stretch S mod legs by phi, and so scale up volume by phi to the 3rd.  s3 (lowercase s) means shrink S mod legs by 1/phi, and so scale down volume by (1/phi) to the 3rd.

In [23]:
(60 * Smod + 20 * Smod * (1/phi**3)).evalf(40)

2.917960675006309107724789938061712936781

In [24]:
IcosaWithin

5*(-sqrt(10) + 3*sqrt(2))**2/2

In [25]:
IcosaWithin.evalf(40)

2.917960675006309107724789938061712936781

## Fun with Classes

Now it's time to do some actual object oriented programming by defining a poly type that may be hit with either a volume scaler, or an edge scalar. These two have a consistent power relationship meaning to rescale volume by X is to rescale edges by ${X}^{1/3}$. To rescale edges by X is to rescale volume by $X^{3}$.


David Koski:

Fundamental to E->S is phi scaling:
*   (1) E = (4) e3 + (1) e6
*   (1) S= (4) s3 + (1) s6

In [26]:
class Poly:

    def __init__(self, spine, vol):
        """
        spine may be any linear dimension we care about
        volume is shape volume
        """
        self.edge = spine
        self.volume = vol

    def __mul__(self, scale_v):
        """
        scale_v resizes volume, and edges by a 3rd root of scale_v
        """
        return type(self)(self.edge * scale_v**(1/3), self.volume * scale_v)

    def __call__(self, scale_e):
        """
        scale_e resizes edges, and volume by a 3rd power of scale_e
        """
        return type(self)(self.edge * scale_e, self.volume * scale_e**3)

In [27]:
rt2   = sy.sqrt       # map sympy sqrt to rt2 for 2nd root
φ     = (rt2(5)+1)/2  # golden ratio -- same as phi above

E  = Poly(1, (rt2(2) / 8) * (φ**-3)) # home base Emod
e3 = E  * φ**-3  # small e, phi down
e6 = e3 * φ**-3  # small e, phi down again
E3 = E  * φ**3   # capital E, phi up
E6 = E3 * φ**3   # phi up yet again

S  = Poly(sfactor, (φ**-5) / 2)   # home base Smod
s3 = S  * φ**-3  # small s, phi down
s6 = s3 * φ**-3  # small s, phi down again
S3 = S  * φ**3   # capital s, phi up
S6 = S3 * φ**3   # phi up yet again

In [28]:
(4 * e3.volume + e6.volume)

sqrt(2)/(8*(1/2 + sqrt(5)/2)**9) + sqrt(2)/(2*(1/2 + sqrt(5)/2)**6)

In [29]:
rt2(2)/(8*φ**9) + rt2(2)/(2*φ**6)

sqrt(2)/(8*(1/2 + sqrt(5)/2)**9) + sqrt(2)/(2*(1/2 + sqrt(5)/2)**6)

In [30]:
(4 * e3.volume + e6.volume).evalf(40)

0.04173131692777365429943951200166529707253

In [31]:
E.volume.evalf(40)

0.04173131692777365429943951200166529707253

In [32]:
S.volume/E.volume

2*sqrt(2)/(1/2 + sqrt(5)/2)**2

In [33]:
(2 * rt2(2))/φ**2

2*sqrt(2)/(1/2 + sqrt(5)/2)**2

In [34]:
((2 * rt2(2))/φ**2).evalf(40)

1.080363026950905814406172628196375701989

In [35]:
2 * rt2(7 - 3 * rt2(5))

2*sqrt(7 - 3*sqrt(5))

In [36]:
2 * rt2(7 - 3 * rt2(5)).evalf(40)

1.080363026950905814406172628196375701989

However, we may also grow or shrink a volume by transforming it to another shape. We think of the CO to Icosa transformation, from volume 20 to 18.51 as shape shifting. The edges didn't resize at all. 

Application of 1/sfactor in this case is not meant to preserve the CO shape.

What we show below is a closed loop where we start from the 2.5 volumed CO(R) embedded in Octa(D). Two applications of the sfactor (i.e. 2nd power) stretches its legs as the VE twist-opens into Icosa(sfactor), faces embedded in Octa(D) all along the way.

An additional shape shifting application of sfactor turns Icosa(sfactor) back into a CO with sfactor edges (i.e. unchanged -- the usual Jitterbug action).

Now that we're at this larger CO(sfactor) of volume about 3.15, one application of 1/sfactor, strictly as a shape-preserving scaling factor this time, will fast track us back, in one step, back to the CO(R) of 2.5.

It's like an express train versus a local. The local scenic route goes from CO(R) through some other intermediate step along the Octa(D) edges rail line -- what the CO(R)'s vertices follow when opening to IcosaWithin (volume ~2.92). 

Yet a 3rd application of sfactor, as a shape shifter, does the classic Icosa --> Cubocta Jitterbug while preserving edge lengths. We're now at our destination: CO(sfactor) i.e. a cuboctahedron with edges (and radii) equal sfactor (S:E).

The express route is CO(sfactor) back to CO(R) direct, with 1/sfactor turning edges of sfactor (~1.08) back to edges of 1R, back to where we started in one step.

In [41]:
CO = Poly(1, 2.5)
IcosaWithin = CO * sfactor * sfactor
IcosaWithin.volume.evalf()

2.91796067500631

In [42]:
COmax = IcosaWithin * sfactor
COmax.volume.evalf()

3.15245682737352

In [44]:
COmax

In [51]:
(COmax(1/sfactor)).volume.evalf()

2.50000000000000

In [53]:
(CO * sfactor * sfactor).volume.evalf()

2.91796067500631

In [54]:
(CO * sfactor).volume.evalf() # somewhere along the Octa(D) rail system

2.70090756737726